### PR TITLE
Dev/1.15.7

### PR DESCRIPTION
**Pull request by alexjhawk**

### CODE REVIEW

File: src/main/java/com/hms_networks/americas/sc/extensions/datapoint/DataPoint.java [hunk 1]
```diff
@@ -133,9 +133,16 @@ public String toString() {
   public abstract DataType getType();
 
   /**
-   * Get the string representation of the data point value.
+   * Get the {@link String} representation of the data point value.
```
Comment (TomKimsey) [inline src/main/java/com/hms_networks/americas/sc/extensions/datapoint/DataPoint.java:136]: This belongs in a separate commit

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+   *
+   * @param values values to sum
+   * @return sum of values
+   * @throws IllegalArgumentException if values are not of the same type
+   * @since 1.15.7
+   */
+  public static Object getValueSum(Object[] values) {
+    // Sum values by type
+    Object sum;
+    if (values[0] instanceof Integer) {
```
Comment (TomKimsey) [inline src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:209]: 0 should be a named variable.

**Review by TomKimsey — CHANGES_REQUESTED** [review]

File: src/main/java/com/hms_networks/americas/sc/extensions/datapoint/DataPoint.java (hunk 1 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/datapoint/DataPoint.java:136]: Changed (and covered the instances of this in subclasses)

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 2 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java]: Moved the entire `values[0]` out to a variable because it is referenced multiple times below.

**Review by TomKimsey — DISMISSED** [review]

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    return minValue;
+  }
+
+  /**
+   * Get and return the maximum value of the two specified values.
+   *
+   * @param value1 first value
+   * @param value2 second value
+   * @return maximum value
+   * @throws IllegalArgumentException if values are not of the same type
```
Comment (it-hms) [inline src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:106]: or are not types that support comparison

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+   * @return sum of values
+   * @throws IllegalArgumentException if values are not of the same type
+   * @since 1.15.7
+   */
+  public static Object getValueAvg(Object[] values) {
+    // Get sum of values
+    Object sum = getValueSum(values);
+
+    // Get average of values
+    Object avg = getValueDivide(sum, values.length);
```
Comment (it-hms) [inline src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:192]: for empty array, divide by zero. Perhaps check and throw nicer exception.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    // Return average
+    return avg;
+  }
+
+  /**
+   * Get and return the sum of the specified values.
+   *
+   * @param values values to sum
+   * @return sum of values
+   * @throws IllegalArgumentException if values are not of the same type
```
Comment (it-hms) [inline src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:203]: I don't think there is any checking if values are the same type.  
If values are not types that support addition? 
Should add comment that caller is responsible values being a homogenous list

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+  /**
+   * Get and return the value of the specified value divided by the specified divisor.
+   *
+   * @param value the value to divide
+   * @param divisor the divisor
+   * @return the value divided by the divisor
+   * @throws IllegalArgumentException if the value is not of a supported type
+   * @since 1.15.7
+   */
+  public static Object getValueDivide(Object value, int divisor) {
```
Comment (it-hms) [inline src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:271]: divisor is an int, so this is meant for average calculation, not really general purpose. Should be private? Or should  divisor be object type?

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+    } else if (firstValue instanceof Short) {
+      // Sum values
+      short sumShort = 0;
+      for (int i = 0; i < values.length; i++) {
+        sumShort += ((Short) values[i]).shortValue();
+      }
+      sum = new Short(sumShort);
+    } else if (firstValue instanceof Byte) {
+      // Sum values
+      byte sumByte = 0;
```
Comment (it-hms) [inline src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:247]: will there be a rollover if sum of bytes>255?  Perhaps sum should always return long or double?

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+
+  /**
+   * Get and return the sum of the specified values.
+   *
+   * @param values values to sum
+   * @return sum of values
+   * @throws IllegalArgumentException if values are not of the same type
+   * @since 1.15.7
+   */
+  public static Object getValueAvg(Object[] values) {
```
Comment (it-hms) [inline src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:187]: This implementation does not work well for Bytes. If byte[] is longer than 255 or sum > 255

**Review by it-hms — CHANGES_REQUESTED** [review]
I don't think getValueAvg will work for Bytes due to roll over. May not work for ints too, but this is less of a concern.  Perhaps needs to throw rollover exception, or needs to perform sum and division with long types.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 8 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:187]: This is a side effect of supporting all of the primitive types in Java, and this class would likely not be necessary on Java 1.5+.

It doesn't make sense for us to add special handling based on specific types. Ultimately, this is a helper method that 'replaces' newer Java code like this:

```
int sum = 0;
for (int x: values) sum += x;
```

Where `sum += x` is the same as `sum = sum + x;`, the same issue could arise here regardless of type. It is a general consideration when summing numbers, and our implementation should handle this exactly the same as using a loop to sum the values in newer Java, without a utility class.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 7 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:247]: Java has underflow/overflow, but it does not explicitly catch or throw an error in that event.

This implementation matches what is normally expected out of newer Java code, where the sum of two integers would return an integer. The JVM does not automatically convert types to work around overflow concerns. For example, in the following Java code, the raw type `T` would be `int` (note: Java does not actually allow raw/parameterized primitive types, this is just an example):

```
int x = 10;
int y = 10;
T sum = x + y;
```

Regardless of the values in `x` and `y`, the type for `sum`, `T` is always going to be `int`. If the integers are too large, it would be expected that they are converted *beforehand* (before being summed, or in this case, passed to the util method). This would look like 

```
int x = 10;
int y = 10;
T sum = (long) x + (long) y;
```

where `T` would in this case be the type `long`.

By changing the return type to be something other than the input type(s), the calling method is likely to run into errors as well, because it would expect the result to be of the same type passed in, and then a class cast exception would occur when trying to convert the returned `Object` to the expected type, because the type does not match.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 4 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:192]: I will add a check. Given the constraints imposed by `getValueSum(...)`, an empty array would cause an error before reaching here, but it doesn't hurt to double-check.

It would be expected that in `getValueSum(...)` since the first value is null, it would not meet any of the conditions such as `instanceof Integer`, `instanceof Double`, etc, thus resulting in an IllegalArgumentException with a message similar to: `Cannot sum values of type null.`

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 7 above)
Comment (it-hms) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:247]: What do you think is a good solution? drop support for Bytes?

Comment (TomKimsey) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:247]: As an alternative, what about creating a OverflowException?

Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:247]: I don't think any changes are required. There is no reason for us to deviate from what the JVM would normally do, and this method is no different from something like
`int a = b + c;` except we are casting the types based on `instanceof` due to Java 1.4 limitations.

Regarding support for 'Byte', this likely won't receive much use but is included so that the utility class maintains support for *all* primitive types.

Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:247]: > As an alternative, what about creating a OverflowException?

This could be done, but at the end of the day, dealing with overflow/underflow is out of the scope here I think.

This method is concerned with casting the types (as detected) and then summing the two number values exactly as they would normally in Java, with no extra processing or handling.

Comment (TomKimsey) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:247]: Im not sure its fair to compare an operator to a method. Using operators you always have the ability to cast. In this method you dont provide that ability. IF you could specify the return type(/type to use for summation) i think that would be more comparable.

**Comment by alexjhawk** [comment]
I am reviewing two customer support requests and will revisit the outstanding changes on this PR later today.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 7 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java:247]: The intention of this method is to act as an operator, though, with the advantage being that it hides a lot of the casting required in Java 1.4 when using primitive values in their corresponding `Object` class, i.e. `Integer`. This includes the side effects, positive or negative, from the values you put in.

I will take a further look at how the JVM handles summation with multiple values and types. If there are improvements that can be made in this method to more closely mimic the expectations/actual JVM behavior (when directly summing), I will change.

The goal here is to be able to have two or more data points from the same tag, take their values from `DataPoint#getValueObject()`, and then get the corresponding min/max/avg values without the large amounts of type detection and casting code at the connector/application level.

**Comment by alexjhawk** [comment]
> I am reviewing two customer support requests and will revisit the outstanding changes on this PR later today.

I am back to working on this PR. I have made a few improvements to the divide method, and have a few other changes upcoming which will better align with the exact JVM behaviors. Alongside this will be a JUnit test class to verify the functionality of these methods.

**Comment by alexjhawk** [comment]
> > I am reviewing two customer support requests and will revisit the outstanding changes on this PR later today.
> 
> I am back to working on this PR. I have made a few improvements to the divide method, and have a few other changes upcoming which will better align with the exact JVM behaviors. Alongside this will be a JUnit test class to verify the functionality of these methods.

I have made several changes to better align with the exact JVM behavior if the operations were performed directly with the primitive values. The JUnit tests I added to go along with this class are complete and fairly extensive, given the assistance of GitHub CoPilot. Unfortunately, while many of these tests are passing, a few fail, so I need to address those discrepancies first.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+ */
+public class RawNumberValueUtils {
+
+  /**
+   * Get and return the minimum value of the two specified values.
+   *
+   * @param value1 first value
+   * @param value2 second value
+   * @return minimum value
+   * @throws IllegalArgumentException if values are not of the same type
```
[the remark this answers is not in the document]
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java]: I've made numerous changes to the Javadocs to more clearly outline the input restrictions and the resulting exceptions should those be violated.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 3 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java]: https://github.com/hms-networks/sc-ewon-flexy-extensions-lib/pull/81#discussion_r1385311719

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 5 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java]: https://github.com/hms-networks/sc-ewon-flexy-extensions-lib/pull/81#discussion_r1385311719

`Values` does not necessarily need to be a list of all the same types, as long as they are all implementations of `Number`. Checking has been added/ensured for this.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 6 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java]: Updated so `divisor` is an `Object`.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 7 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java]: A method overload was added to allow for overflow handling of the integer type. This covers byte, short, and int. If an overflow is detected and the option is true, the returned result is promoted to `Long` so it contains the true result. This is in addition to supporting the primitive operation behavior where the overflow would still return an integer result.

File: src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java (hunk 8 above)
Comment (alexjhawk) [reply src/main/java/com/hms_networks/americas/sc/extensions/util/RawNumberValueUtils.java]: Java actually does do promotion from 'byte' or 'short' to an 'int' when performing operations on the values. I have updated all applicable methods/code to account for this. The JUnit test class also verifies this behavior.

**Comment by alexjhawk** [comment]
All JUnit test failures have been resolved and PR is ready for review again. 

A majority of the failures were associated with the improper handling of short/byte operations, and the remaining few were resolved by correcting tests to account for the operation results for byte/short types to be int.

Basically, Java does not handle overflow for the corresponding primitive operations, but it does promote byte/short to int when performing operations, which resolves overflow issues/concerns for those two types. This has been changed/reflected in the code now, and JUnit tests have been added to verify compliance with all of the supported types.

**Review by it-hms — DISMISSED** [review]

File: web-docs/docs/02-CHANGELOG.mdx [hunk 10]
```diff
@@ -23,7 +32,8 @@ toc_max_heading_level: 2
 
 ## Version 1.15.3
 ### Features
-- Added service timer to the application reboot watchdog to ensure application is in a stable state before allowing the watchdog to be serviced.
+- Added service timer to the application reboot watchdog to ensure application is in a stable state
```
Comment (TomKimsey) [inline web-docs/docs/02-CHANGELOG.mdx:35]: A lot of previous versions are getting touched here. These do not belong in this commit. Also is there the intention to add a max line length parameter to .mdx files?

**Review by TomKimsey — CHANGES_REQUESTED** [review]

File: web-docs/docs/02-CHANGELOG.mdx (hunk 10 above)
Comment (alexjhawk) [reply web-docs/docs/02-CHANGELOG.mdx:35]: I will correct this. I didn't intend to establish a maximum length in MDX files, although IntelliJ formats according to the same max line length as our Java code. I just missed the changes leaking into the release commit.

That being said, any new MD/MDX I create generally does adhere to the same max line length requirement as our Java code.

Comment (alexjhawk) [reply web-docs/docs/02-CHANGELOG.mdx]: Fixed

**Review by TomKimsey — APPROVED** [review]
